processor/lsmintervalprocessor/internal/merger/value.go

```diff
@@ -0,0 +1,796 @@
+	}
+	return scopeID, nil
+}
+
+// addMetric adds the given metric to the store while also considering
+// datapoint limiters. If a limit is configured and breached by adding a new
+// metric then the datapoint overflow is updated and the metric is discarded
+// as when datapoint overflows, a new metric overflow sum metric is added
+// with delta temporality tracking the cardinality estimate of the overflow.
+func (s *Value) addMetric(
+	scopeID identity.Scope,
+	otherMetric pmetric.Metric,
+) identity.Metric {
+	metricID := identity.OfMetric(scopeID, otherMetric)
+	if _, ok := s.metricLookup[metricID]; ok {
+		return metricID
+	}
+	scope := s.scopeLookup[scopeID]
+
+	// Metrics doesn't have overflows (only datapoints have)
+	// Clone it *without* the datapoint data
+	m := scope.Metrics().AppendEmpty()
+	m.SetName(otherMetric.Name())
+	m.SetDescription(otherMetric.Description())
+	m.SetUnit(otherMetric.Unit())
+	switch otherMetric.Type() {
+	case pmetric.MetricTypeGauge:
+		m.SetEmptyGauge()
+	case pmetric.MetricTypeSummary:
+		m.SetEmptySummary()
+	case pmetric.MetricTypeSum:
+		otherSum := otherMetric.Sum()
+
+		sum := m.SetEmptySum()
+		sum.SetAggregationTemporality(otherSum.AggregationTemporality())
+		sum.SetIsMonotonic(otherSum.IsMonotonic())
+	case pmetric.MetricTypeHistogram:
+		otherHist := otherMetric.Histogram()
+
+		hist := m.SetEmptyHistogram()
+		hist.SetAggregationTemporality(otherHist.AggregationTemporality())
+	case pmetric.MetricTypeExponentialHistogram:
+		otherExp := otherMetric.ExponentialHistogram()
+
+		exp := m.SetEmptyExponentialHistogram()
+		exp.SetAggregationTemporality(otherExp.AggregationTemporality())
+	}
+	s.metricLookup[metricID] = m
+	return metricID
+}
+
+// addSumDataPoint returns a data point entry in the store for the given metric
+// and the external data point if it is present. If the data point is not
+// present then either a new data point is added or if the data point overflows
+// due to configured limit then an empty data point is returned. The returned
+// bool value is `true` if a new data point is created and `false` otherwise.
+func (s *Value) addSumDataPoint(
+	metricID identity.Metric,
+	otherDP pmetric.NumberDataPoint,
+) (pmetric.NumberDataPoint, bool) {
+	streamID := identity.OfStream(metricID, otherDP)
+	if dp, ok := s.numberLookup[streamID]; ok {
+		return dp, false
+	}
+	sm := s.scopeLookup[metricID.Scope()]
+	metric := s.metricLookup[metricID]
+	if sm.datapointsLimits.CheckOverflow(metricID.Hash().Sum64()) {
+		// Datapoints overflow detected. In this case no action has to be
+		// done at this point since data point overflow should create a new
+		// overflow metric of sum type recording the number of unique
+		// datapoints. This number will be recorded in the limit tracker
+		// and the metric will be populated on demand.
+		return pmetric.NumberDataPoint{}, false
+	}
+	dp := metric.Sum().DataPoints().AppendEmpty()
+	s.numberLookup[streamID] = dp
+	return dp, true
```

Comment:
Perhaps the caller could be simplified if we called `otherDP.CopyTo(dp)` here? Then instead of returning true here, we would return true only when the caller should update the returned data point. WDYT? (Same for other metric types of course.)